charms/worker/k8s/src/charm.py

```
@@ -1061,14 +1106,62 @@ def _get_external_kubeconfig(self, event: ops.ActionEvent):
         try:
             server = event.params.get("server")
             if not server:
-                log.info("No server requested, use public-address")
-                server = f"{_get_public_address()}:6443"
+                log.info("No server requested, use public address")
+                server = self._get_public_address()
+                if not server:
+                    event.fail("Failed to get public address. Check logs for details.")
+                    return
+                log.info("Found public address: %s", server)
+                port = str(APISERVER_PORT)
+                if self.is_control_plane and self.external_load_balancer.is_available:
+                    log.info("Using external load balancer port as the public port")
+                    port = str(EXTERNAL_LOAD_BALANCER_PORT)
```

it feels like this should be a new method called `_get_public_endpoint()`.  It should return the "most public" endpoint available for the charm.  the order should be:

```
https://<external_load_balancer-ip>:LB_PORT
https://<leader CP unit ingress address>:INTERNAL_PORT
https://<other CP units ingress addresses>:INTERNAL_PORT
https://127.0.0.1:INTERNAL_PORT
```

it should ALWAYS arrive at SOMETHING if k8s is bootstrapped.

---

> it should ALWAYS arrive at SOMETHING if k8s is bootstrapped.

That's right but at the same time if we have an external load balancer and we fail to retrieve its address, I think it's best to fail rather than returning the second best option (which is probably not so useful, we have a loadbalancer and we're probably don't want to talk directly to our backends).